### PR TITLE
Various edits for correctness

### PR DESCRIPTION
**Pull request by littledan**

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -478,7 +478,8 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
             1. <ins>Set _module_.[[PendingAsyncDependencies]] to _module_.[[PendingAsyncDependencies]] + 1.</ins>
             1. <ins>Append _module_ to _requiredModule.[[AsyncParentModules]]_.</ins>
         1. <ins>If _module_.[[PendingAsyncDependencies]] is 0, then</ins>
-          1. Perform ? _module_.ExecuteModule()
+          1. Perform ! ExecuteCyclicModule(_module_).
+          1. <ins>If _module_.[[EvaluationError]] is not *undefined*, return ThrowCompletion(_module_.[[EvaluationError]].</ins>
```
Comment (guybedford) [inline spec.html]: Missing closing bracket - `return ThrowCompletion(_module_.[[EvaluationError]])`.

File: spec.html [hunk 2]
```diff
@@ -528,26 +559,27 @@ <h1><ins>AsyncExecutionFulfilled ( _module_ )</ins></h1>
             1. Assert _m_.[[Status]] is `"evaluating-async"`.
             1. Let _cycleRoot_ be ! GetCycleRoot(_m_).
             1. If _cycleRoot_.[[EvaluationError]] is not *undefined*, return *undefined*.
-            1. Perform ! _module_.ExecuteModule().
+            1. Perform ? ExecuteCyclicModule(_module_).
```
Comment (guybedford) [inline spec.html]: Should be `!`.

File: spec.html [hunk 3]
```diff
@@ -514,12 +544,13 @@ <h1><ins>GetCycleRoot ( _module_ )</ins></h1>
...
-      <h1><ins>AsyncExecutionFulfilled ( _module_ )</ins></h1>
+    <emu-clause id="sec-cyclicmoduleexecutionfulfilled" aoid="CyclicModuleExecutionFulfilled">
+      <h1><ins>CyclicModuleExecutionFulfilled ( _module_ )</ins></h1>
       <emu-alg>
-        1. Assert: _module_.[[Status]] is `"evaluating-async"`.
         1. Assert: _module_.[[EvaluationError]] is *undefined*.
-        1. Set _module_.[[Status]] to `"evaluated"`.
+        1. If _module_.[[Async]] is *true*,
+          1. Assert: _module_.[[Status]] is `"evaluating-async"`.
+          1. Set _module_.[[Status]] to `"evaluated"`.
```
Comment (guybedford) [inline spec.html:556]: A sync parent of an async module wouldn't be transitioned to "evaluated" with this guard.

I think this needs to be:

```
1. Assert _module_.[[EvaluationError]] is *undefined*.
1. If _module_.[[Status]] is `"instantiating"`, then
  1. Assert: _module_.[[Async]] is *false*.
  1. Assert: _module_.[[AsyncParentModules]] is an empty List.
1. If _module_.[[Status]] is `"evaluating-async"`, then
  1. Set _module_.[[Status]] to `"evaluated"`.
```

File: spec.html [hunk 4]
```diff
@@ -528,26 +559,27 @@ <h1><ins>AsyncExecutionFulfilled ( _module_ )</ins></h1>
...
-      <h1><ins>AsyncExecutionRejected ( _module_, _error_ )</ins></h1>
+    <emu-clause id="sec-cyclicmoduleexecutionrejected" aoid="CyclicModuleExecutionRejected">
+      <h1><ins>CyclicModuleExecutionRejected ( _module_, _error_ )</ins></h1>
       <emu-alg>
-        1. Assert: _module_.[[Status]] is `"evaluating-async"`.
         1. Assert: _module_.[[EvaluationError]] is *undefined*.
-        1. Set _module_.[[Status]] to `"evaluated"`.
+        1. If _module_.[[Async]] is *true*,
+          1. Assert: _module_.[[Status]] is `"evaluating-async"`.
+          1. Set _module_.[[Status]] to `"evaluated"`.
```
Comment (guybedford) [inline spec.html:582]: Same comment as with the `CyclicModuleExecutionFulfilled` case.

**Review by guybedford — COMMENTED** [review]
I finally see what you mean about the sync completions not being attached. Well spotted on that, this looks like a great reworking.

The only issue remaining algorithmically is to fix up the sync state transitions in the completion handlers, which I've commented on.

It all works out really elegantly though otherwise, thanks.

File: spec.html [hunk 5]
```diff
@@ -501,6 +502,35 @@ <h1>InnerModuleEvaluation( _module_, _stack_, _index_ )</h1>
...
 
+    <emu-clause id="sec-execute-cyclic-module" aoid="ExecuteCyclicModule">
+      <h1><ins>ExecuteCyclicModule ( _module_ )</ins></h1>
+      <emu-alg>
+        1. If _module_.[[Async]] is *false*, then:
+          1. Let _result_ be _module_.ExecuteModule().
+          1. If _result_ is a normal completion,
+            1. Perform ! CyclicModuleExecutionFulfilled(_module_).
+          1. Otherwise,
+            1. Perform ! CyclicModuleExecutionRejected(_module_, _result_.[[Value]])
```
Comment (guybedford) [inline spec.html]: Missing `.`

File: spec.html (hunk 3 above)
Comment (littledan) [reply spec.html:556]: Good point! Fixing.

**Comment by littledan** [comment]
Merged upstream.

**Pull request closed without merging**